apps/prefixSum/prefixSum-stdexec.cpp

```diff
@@ -0,0 +1,209 @@
+ * SOFTWARE.
+ */
+
+/*
+ * commons for the prefixSum codes
+ */
+
+#define PSUM_STDEXEC
+#include "prefixSum.hpp"
+#include "repeat_n/repeat_n.cuh"
+
+//
+// stdexec prefixSum function
+//
+template <typename T>
+[[nodiscard]] ull_t* prefixSum(scheduler auto &&sch, const T *in, const int N)
+{
+    ull_t *y = new ull_t[N+1];
+
+    // memcpy to output vector
+    ex::sync_wait(ex::schedule(sch) | ex::bulk(N, [=](int k){ y[k] = in[k]; }));
+
+    int niters = ilog2(N);
+    int *d_ptr = new int(0);
+
+    // GE Blelloch (1990) algorithm
+    // upsweep algorithm
+    ex::sender auto uSweep = ex::just()
+    | exec::on(sch,
+       repeat_n(niters, ex::bulk(N/2, [=](int k){
+            int d = *d_ptr;
+            int s1 = 1 << d+1;
+            int s2 = 1 << d;
+            int my = (k+1) * s1 - 1;
+            bool isActive = (k < N/s1);
+
+            // only update the participating indices
+            if (isActive)
+                y[my] += y[my-s2];
+        })
+        | ex::then([=](){ *d_ptr += 1; })
+    ));
+
+    ex::sync_wait(uSweep);
+
+    // write sum to y[N] and reset vars
+    ex::sync_wait(schedule(sch) | ex::then([=](){
+        y[N] = y[N-1];
+        y[N-1] = 0;
+        *d_ptr = niters-1;
+    }));
+
+    // downsweep algorithm
+    ex::sender auto dSweep = ex::just()
+    | exec::on(sch,
+       repeat_n(niters, ex::bulk(N/2, [=](int k){
+            int d = *d_ptr;
+            int s1 = 1 << d+1;
+            int s2 = 1 << d;
+            int my = (k+1) * s1 - 1;
+            bool isActive = (k < N/s1);
+
+            // only update the participating indices
+            if (isActive)
+            {
+                auto tmp = y[my];
+                y[my] += y[my-s2];
+                y[my-s2] = tmp;
+            }
+
+        })
+        | ex::then([=](){ *d_ptr -= 1; })
+    ));
+
+    ex::sync_wait(dSweep);
+
+    return y;
+}
+
+//
+// simulation
+//
+int main(int argc, char* argv[])
+{
+    // parse params
+    const prefixSum_params_t args = argparse::parse<prefixSum_params_t>(argc, argv);
+
+    // see if help wanted
+    if (args.help)
+    {
+        args.print();  // prints all variables
+        return 0;
+    }
+
+    // simulation variables
+    int N = args.N;
+    bool print_arr = args.print_arr;
+    bool print_time = args.print_time;
+    bool validate = args.validate;
+    std::string sched = args.sch;
+    int max_threads = args.max_threads;
+
+    if (!isPowOf2(N))
+    {
+        N = ceilPowOf2(N);
+        std::cout << "INFO: N != pow(2). Setting => N = " << N << std::endl;
+    }
+
+    // input data
+    data_t *in = new data_t[N];
+
+    // random number generator
+    std::random_device rd;
+    std::mt19937 gen(rd());
+    std::uniform_int_distribution<data_t> dist(1, 10);
+
+    std::cout << "Progress:0%" << std::flush;
+
+    // fill random between 1 to 10
+    std::generate(in, in+N, [&]() { return dist(gen); });
```

Comment:
should this initialization process be part of benchmark timing?

Reply:
also, should we use parallel execution policy here as well?

Reply:
it's not a part of the benchmark aka timer, only prints progress marks

Reply:
RNG (random number generators) don't behave well in multithreaded mode. There are ways to generate RNG in parallel but not really needed here. Using `std::execution::par` would not compile here.